### DISCUSSION TITLE
Removing backgrounds from a `NavigationSplitView`

### DISCUSSION THREAD

**Discussion by slaght** [siteline/SwiftUI-Introspect#292] [Q&A] [answered]
Hey all,

Firstly wanted to say thanks for your work on the library, it's proven super useful so far.

I'm using `SwiftUIIntrospect 0.8.0` on an iOS 17 project to try to remove the background from a `NavigationSplitView` (I am displaying a view underneath it and want it to persist throughout the application). The following code is working for removing the backgrounds initially: 

```
...
NavigationSplitView(columnVisibility: NavigationSplitViewVisibility.detailOnly) {
	List(menuItems, selection: $selectedItem) { item in
		Text(item.name)
	}
} detail: {
	TestView()
}
.introspect(.navigationSplitView, on: .iOS(.v17)) {
	$0.viewControllers.forEach { controller in
		controller.parent?.view.backgroundColor = .clear
		clearBackgrounds(view: controller.view)
	}
}
...

func clearBackgrounds(view: UIView) {
	view.backgroundColor = .clear
	for subview in view.subviews {
		clearBackgrounds(view: subview)
	}
}
```

This works beautifully... until a detail view gets presented when clicking on a list item. When this happens it appears that the backgrounds reappear: 

![Screenshot 2023-07-02 at 6 57 10 PM](https://github.com/siteline/SwiftUI-Introspect/assets/7023112/135c05c0-ff67-4342-92c1-86ac096a1314)

How do I introspect the detail views to remove their backgrounds as well? I have tried introspecting NavigationSplitView in the `TestView` instance in the detail declaration but that doesn't seem to do anything.

**Comment by davdroman** [comment, marked as answer]
Thanks for your question, @slaght.

Indeed, setting `backgroundColor`s via introspection is flakey. SwiftUI insists on overriding the background color of its UIKit components on each state update. However, I'm not convinced this is necessarily a problem begging for introspection, as nowadays SwiftUI offers modifiers to clear out a number of backgrounds. Instead, a hybrid approach might be best.

Here's what I came up with:

https://github.com/davdroman/NavigationSplitViewRemoveBackgrounds

Note I took the liberty to drop in a `.navigationTransition(.slide)` modifier from my https://github.com/davdroman/swiftui-navigation-transitions library to make the transition smoother, because the default iOS transition looks quite weird with transparent backgrounds on.

Let me know what you think.

**Reply by slaght** [reply]
Wow, thanks for taking the time to create an entire sample project, this works perfectly. Managed to adapt it to my project and make it a little more versatile by checking for and ignoring views of certain types (ex. ListCollectionViewCell) that should maintain their backgrounds. Will also be taking a look at your navigation transition library because I can see the wonky transition effects you mentioned.
